Mathlib/Analysis/Calculus/Rademacher.lean

```diff
@@ -0,0 +1,404 @@
+
+## Implementation
+
+There are many proofs of Rademacher's theorem. We follow the one by Morrey, which is not the most
+elementary but maybe the most elegant once necessary prerequisites are set up.
+* Step 0: without loss of generality, one may assume that `f` is real-valued.
+* Step 1: Since a one-dimensional Lipschitz function has bounded variation, it is differentiable
+almost everywhere. With a Fubini argument, it follows that given any vector `v` then `f` is ae
+differentiable in the direction of `v`. See `LipschitzWith.ae_lineDifferentiableAt`.
+* Step 2: the line derivative `LineDeriv ℝ f x v` is ae linear in `v`. Morrey proves this by a
+duality argument, integrating against a smooth compactly supported function `g`, passing the
+derivative to `g` by integration by parts, and using the linearity of the derivative of `g`.
+See `LipschitzWith.ae_lineDeriv_sum_eq`.
+* Step 3: consider a countable dense set `s` of directions. Almost everywhere, the function `f`
+is line-differentiable in all these directions and the line derivative is linear. Approximating
+any direction by a direction in `s` and using the fact that `f` is Lipschitz to control the error,
+it follows that `f` is Fréchet-differentiable at these points.
+See `LipschitzWith.hasFderivAt_of_hasLineDerivAt_of_closure`.
+-/
+
+open Filter MeasureTheory Measure FiniteDimensional Metric Set Asymptotics
+
+open scoped BigOperators NNReal ENNReal Topology
+
+variable {E : Type*} [NormedAddCommGroup E] [NormedSpace ℝ E] [FiniteDimensional ℝ E]
+  [MeasurableSpace E] [BorelSpace E]
+  {F : Type*} [NormedAddCommGroup F] [NormedSpace ℝ F] {C D : ℝ≥0} {f g : E → ℝ} {s : Set E}
+  {μ : Measure E} [IsAddHaarMeasure μ]
+
+namespace LipschitzWith
+
+/-!
+### Step 1: A Lipschitz function is ae differentiable in any given direction
+
+This follows from the one-dimensional result that a Lipschitz function on `ℝ` has bounded
+variation, and is therefore ae differentiable, together with a Fubini argument.
+-/
+
+theorem ae_lineDifferentiableAt_of_prod
+    {C : ℝ≥0} {f : E × ℝ → ℝ} (hf : LipschitzWith C f) {μ : Measure E} :
+    ∀ᵐ p ∂(μ.prod volume), LineDifferentiableAt ℝ f p (0, 1) := by
+  apply (ae_prod_mem_iff_ae_ae_mem (measurableSet_lineDifferentiableAt hf.continuous)).2
+  apply eventually_of_forall (fun x ↦ ?_)
+  have : ∀ᵐ (y : ℝ), DifferentiableAt ℝ (fun z ↦ f (x, z)) y := by
+    apply LipschitzWith.ae_differentiableAt_real (C := C)
+    intro z z'
+    convert hf (x, z) (x, z')
+    simp [Prod.edist_eq]
+  filter_upwards [this] with y hy
+  have h'y : DifferentiableAt ℝ (fun z ↦ f (x, z)) (y + 0) := by simpa using hy
+  have : DifferentiableAt ℝ (fun t ↦ y + t) 0 :=
+    (differentiable_id.const_add _).differentiableAt
+  simpa only [LineDifferentiableAt, Prod.smul_mk, smul_zero, smul_eq_mul, mul_one, Prod.mk_add_mk,
+    add_zero] using h'y.comp 0 this
+
+theorem ae_lineDifferentiableAt (hf : LipschitzWith C f) (v : E) :
+    ∀ᵐ p ∂μ, LineDifferentiableAt ℝ f p v := by
```

Comment:
Does it make sense to have a generic theorem "if a predicate is true a.e. on every `k`-dimensional (`1`-dimensional?) subspace, then it's true a.e.-everywhere w.r.t. a Haar measure"?

Reply:
See #7252